src/cmdlib.sh

```diff
@@ -102,3 +113,81 @@ runcompose() {
          ${TREECOMPOSE_FLAGS:-} ${manifest} "$@"
     set +x
 }
+
+prepare_vm() {
+    # rpms to create appliance VM out of
+    rpms=' bash vim-minimal coreutils util-linux procps-ng kmod kernel-modules'
+    rpms+=' cifs-utils' # for samba
+    rpms+=' systemd' # for clean reboot
+    rpms+=' dhcp-client bind-export-libs iproute' # networking
+    rpms+=' rpm-ostree distribution-gpg-keys' # to run the compose
+    rpms+=' selinux-policy selinux-policy-targeted policycoreutils' #selinux
+
+    # prepare appliance VM
+    supermin -v --prepare --use-installed $rpms -o "${superminpreparedir}"
+    supermin -v --build "${superminpreparedir}" \
+             --include-packagelist --size 4G -f ext2 -o "${superminbuilddir}"
+
+    # Create a disk image for our caches (pkgcache and bare-user build repo)
+    [ ! -d "${workdir}/tmp/emptydir" ] && mkdir "${workdir}/tmp/emptydir"
+    if [ ! -f "${cachesimg}" ]; then
+        virt-make-fs --format=qcow2 --type=xfs \
+                     --size=10G "${workdir}/tmp/emptydir" "${cachesimg}"
+    fi
+}
+
+run_vm() {
+    init=$1
+    set -x
+
+    # Copy in the init script with the code we want to run  
+    chmod +x "${init}"
+    virt-copy-in -a "${superminbuilddir}/root" "${init}" /
+
+    # Execute unprivileged VM to run compose. Some notes:
+    # - smb="${workdir}" - we'll serve our workdir over samba
+    # - sda - rootfs from supermin build
+    # - sdb - caches.qcow2 - we store the bare-user repo and pkgcache here
+    # - discard=unmap - using virtio-scsi disks and discard=unmap so we can fstrim
+    #                   and recover disk space from the VM
+    qemu-kvm -nodefaults -nographic -m 2048 -no-reboot \
+              -kernel "${superminbuilddir}/kernel" \
+              -initrd "${superminbuilddir}/initrd" \
+              -netdev user,id=eth0,hostname=supermin,smb="${workdir}",hostfwd=tcp:127.0.0.1:8000-:8000 \
+              -device virtio-net-pci,netdev=eth0 \
+              -device virtio-scsi-pci,id=scsi0,bus=pci.0,addr=0x3 \
+              -drive if=none,id=drive-scsi0-0-0-0,snapshot=on,file="${superminbuilddir}/root" \
+              -device scsi-hd,bus=scsi0.0,channel=0,scsi-id=0,lun=0,drive=drive-scsi0-0-0-0,id=scsi0-0-0-0,bootindex=1 \
+              -drive if=none,id=drive-scsi0-0-0-1,discard=unmap,file="${cachesimg}" \
+              -device scsi-hd,bus=scsi0.0,channel=0,scsi-id=0,lun=1,drive=drive-scsi0-0-0-1,id=scsi0-0-0-1 \
```

Comment:
Where did you get this list of arguments?

Reply:
handcrafted. some of it I copied from other files in this repo. the scsi lines are there so I can fstrim inside the VM and have it provide free space back to the host. I copied those by setting up a libvirt VM with `virtio-scsi` and `discard=unmap` and then inspecting the qemu arguments..

see: https://dustymabe.com/2013/06/11/recover-space-from-vm-disk-images-by-using-discardfstrim/